lib/gilmour/backends/redis.rb

```diff
@@ -164,13 +169,61 @@ def confirm_subscriber(dest, &blk)
         blk.call(num.to_i > 0)
       end
     rescue Exception => e
-      GLogger.debug e.message
-      GLogger.debug e.backtrace
+      GLogger.error e.message
+      GLogger.error e.backtrace
     end
 
     def stop
       @subscriber.close_connection
     end
 
+    # TODO: Health checks currently use Redis to keep keys in a data structure.
+    # An alternate approach would be that monitor subscribes to a topic
+    # and records nodenames that request to be monitored. The publish method
+    # should fail if there is no definite health monitor listening. However,
+    # that would require the health node to be running at all points of time
+    # before a Gilmour server starts up. To circumvent this dependency, till
+    # monitor is stable enough, use Redis to save/share these data structures.
+    #
+    def register_health_check
+      @publisher.hset RedisHealthKey, self.ident, 'active'
+
+      # - Start listening on a dyanmic topic that Health Monitor can publish
+      # on.
+      #
+      # NOTE: The subscriber is responsible for returning data within 5
+      # seconds, to be considered as healthy, else the node is marked as
+      # unhealthy. 5 consecutive health check failures are marked as permanent
+      # failure and an event is raised for system administrators.
+      #
+      # NOTE: Health checks are not run as forks, to ensure that event-machine's
+      # ThreadPool has sufficient resources to handle new requests.
+      #
+      topic = "gilmour.health.#{self.ident}"
+      add_listener(topic) do
+        respond "Pong"
+      end
+
+      # TODO: Need to do these manually. Alternate is to return the handler
+      # hash from add_listener.
+      @subscriptions[topic][0][:exclusive] = true
+      @subscriptions[topic][0][:timeout] = 5
```

Comment:
All that the handler is doing is calling respond. Any delays in sending back the health would be a result of not having capacity to start the request anyways. So this timeout of 5 seconds is a bit misleading, especially given the comment above.